_posts/2024-03-20-strimzi-kraft-migration.md

```diff
@@ -0,0 +1,233 @@
+my-cluster-zookeeper-0                        1/1     Running   0          20m
+my-cluster-zookeeper-1                        1/1     Running   0          20m
+my-cluster-zookeeper-2                        1/1     Running   0          20m
+strimzi-cluster-operator-7ddf57685d-245pd     1/1     Running   0          17m
+```
+
+#### From ZooKeeper to full KRaft based brokers
+
+The first step is to deploy a `KafkaNodePool` custom resource which hosts the KRaft controllers and then changing the `strimzi.io/kraft` annotation from `disabled` to `migration` to trigger the process.
+
+Following a snippet of a `KafkaNodePool` custom resource to be deployed for getting the KRaft controllers running.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaNodePool
+metadata:
+  name: controller
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  replicas: 3
+  roles:
+    - controller
+  storage:
+    type: jbod
+    volumes:
+      - id: 0
+        type: persistent-claim
+        size: 100Gi
+        deleteClaim: false
+```
+
+The corresponding `Kafka` custom resource is then updated in regards to the `strimzi.io/kraft` set to `migration`. 
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  annotations:
+    strimzi.io/kraft: migration # updated from "disabled" to "migration" to trigger the migration process
+    strimzi.io/node-pools: enabled
+# ...
+```
+
+The annotation change triggers the Strimzi operator to run the following steps for you:
+
+* deploy the KRaft controllers, which are configured with all the connection details to ZooKeeper and with the migration flag enabled;
+* configure the running brokers with the connection details to the KRaft controller quorum, the migration flag enabled and roll them;
+* checking the status of the migration, on each reconciliation, and updating the `Kafka` custom resource to provide such status to the user through the `status.kafkaMetadataState` field;
+* when the migration is done, the brokers are reconfigured to be not connected to ZooKeeper anymore, with the migration flag disabled and they are rolled again.
+
+As you can see, just applying one single annotation value has covered most of the phases in the migration procedure with the user doing actually nothing but following what the operator is running for them.
+
+While the migration is going on, you can see the cluster changing the metadata state during the migration process from `ZooKeeper` to `KRaftPostMigration`, and then waiting for the user to finalize it.
+
+```shell
+NAME         DESIRED KAFKA REPLICAS   DESIRED ZK REPLICAS   READY   METADATA STATE   WARNINGS
+my-cluster   3                        3                     True    ZooKeeper        
+my-cluster   3                        3                     True    ZooKeeper        
+my-cluster   3                        3                     True    ZooKeeper        True
```

Comment:
Should we explain what these warnings mean?

Reply:
I had this in mind but it could mean going into much details about specifying thet IBP and LMF are specified so you have to remove them and so on.

Reply:
I think it's better to spend some words, otherwise the user will be confused.

Reply:
I added a few lines about it.

Reply:
Thanks